### ISSUE TITLE
Provision in session maintenance

### ISSUE THREAD

**Issue by saravanakumarvijayakumar** [hierynomus/sshj#584]
Is there any method available for the following scenario?

1. Connect to SFTP Server through sshj implementation and upload some files.
2. Next time we need to upload one file. It should use the existing session and upload the file. 
The session created in first step should maintain for the particular time. The next upload files should just re-use the session. If there is no upload in the particular period, the session will be closed. The next upload should connect, authenticate and reuse the session.

**Comment by hierynomus** [comment]
You will need to program this yourself. SSHJ does not by itself handle the
reuse for you.

Op vr 22 mei 2020 om 12:42 schreef saravanakumarvijayakumar <
notifications@github.com>:

**Comment by saravanakumarvijayakumar** [comment]
Is there any provision for setting the idle timeout?

**Comment by saravanakumarvijayakumar** [comment]
This is the gentle reminder for the above question.

**Comment by hierynomus** [comment]
@saravanakumarvijayakumar No there isn't. There is a keep alive mechanism available to keep connections (and thus sessions) alive.

**Comment by FabianHenneke** [comment]
I'm trying to do something similar as I would like to use `exec` to execute multiple commands sequentially over a single SSH session. Is this something that is supported or would this require changes to SSHJ code?

**Comment by hierynomus** [comment]
@FabianHenneke You can open a shell for that :) Unless you don't know the commands beforehand, in which case I would suggest to cache the connected client by hostname.

**Comment by FabianHenneke** [comment]
> @FabianHenneke You can open a shell for that :) Unless you don't know the commands beforehand, in which case I would suggest to cache the connected client by hostname.

I guess opening a shell would be a decent workaround, but I would try to avoid that if possible. I may be doing something wrong, but the following attempt at reusing a connected client leads to a "multiple sessions" error:

```Java
SSHClient client = new SSHClient();
client.connect(...);
client.auth(...);
Session session = client.startSession();
session.exec(...);
session.close(...);
client.startSession(); // throws an exception
```

Is there a way to start a second session after closing the first one?

**Comment by hierynomus** [comment]
No, at this moment the SSHClient is not reusable, which is also documented in the JavaDoc.

**Comment by FabianHenneke** [comment]
I took the JavaDoc to mean that a `Session` is not reusable, but if it is indeed the `SSHClient` that is not reusable, this fully explains why I am seeing the exception. I will try to use the shell then.

**Comment by FabianHenneke** [comment]
> I took the JavaDoc to mean that a `Session` is not reusable, but if it is indeed the `SSHClient` that is not reusable, this fully explains why I am seeing the exception. I will try to use the shell then.

As it turns out, `SSHClient` (but *not* `Session`) seems to support reuse: https://github.com/android-password-store/Android-Password-Store/pull/1012
